EasyLM/models/llama/llama_train_ppo.py

```diff
@@ -0,0 +1,605 @@
+    cont_ref_logps = jnp.take_along_axis(jax.nn.log_softmax(cont_ref_logits, axis=-1), cont_input_ids[:, :, None], axis=-1).squeeze(-1) # (B, CL)
+    cont_ref_logps = jax.lax.stop_gradient(cont_ref_logps)
+    timing['time/ppo/reference_forward_pass'] = time.time() - t
+
+    # run forward pass on value
+    t = time.time()
+    cont_values = value_model(input_ids, attn_mask, params=value_train_state.params['params'], dropout_rng=rng_generator()).logits[:, PL-1:-1] # (B, CL)
+    cont_values = jax.lax.stop_gradient(cont_values)
+    timing['time/ppo/value_forward_pass'] = time.time() - t
+
+    # penalize rewards
+    t = time.time()
+    kl = cont_logps - cont_ref_logps # (B, CL)
+    non_score_rewards = -FLAGS.kl_coef * kl # (B, CL)
+    cont_last_token_index = jnp.argmax(cont_position_ids, axis=1) # (B)
+    rewards = non_score_rewards.at[jnp.arange(input_ids.shape[0]), cont_last_token_index].add(score) # (B, CL)
+    rewards = jax.lax.stop_gradient(rewards)
+    timing['time/ppo/compute_rewards'] = time.time() - t
+
+    # compute advantages
+    t = time.time()
+    advantages, returns = compute_advantages(cont_values, rewards, cont_attn_mask) # (B, CL), (B, CL)
+    advantages = jax.lax.stop_gradient(advantages)
+    returns = jax.lax.stop_gradient(returns)
+    timing['time/ppo/compute_advantages'] = time.time() - t
+
+    t = time.time()
+    all_stats = []
+    for ppo_epoch in range(FLAGS.ppo_epochs):
+        assert cont_input_ids.shape[0] % FLAGS.mini_batch_size == 0
+        for mb_start in range(0, cont_input_ids.shape[0], FLAGS.mini_batch_size):
+            mb_end = mb_start + FLAGS.mini_batch_size
+            mb_input_ids = input_ids[mb_start:mb_end]
+            mb_attn_mask = attn_mask[mb_start:mb_end]
+            mb_cont_input_ids = cont_input_ids[mb_start:mb_end]
+            mb_cont_attn_mask = cont_attn_mask[mb_start:mb_end]
+            mb_cont_logps = cont_logps[mb_start:mb_end]
+            mb_cont_values = cont_values[mb_start:mb_end]
+            mb_advantages = advantages[mb_start:mb_end]
+            mb_returns = returns[mb_start:mb_end]
+
+            loss_fn = lambda policy_params, value_params: ppo_loss(
+                policy_model, value_model,
+                policy_params, value_params,
+                rng_generator(),
+                mb_input_ids, mb_attn_mask, mb_cont_input_ids, mb_cont_attn_mask, mb_cont_logps, mb_cont_values, mb_advantages, mb_returns,
+            )
+            grad_fn = jax.value_and_grad(loss_fn, argnums=[0, 1], has_aux=True)
+            (_, stats), (policy_grads, value_grads) = grad_fn(policy_train_state.params, value_train_state.params)
+            policy_train_state = policy_train_state.apply_gradients(grads=policy_grads)
+            value_train_state = value_train_state.apply_gradients(grads=value_grads)
```

Comment:
Do you change the policy/value parameters after each mini_batch? I think the reason why I used mini_batch_size for training and a bigger batch_size for getting rollouts in fgrlhf was to fill out all gpu usage bubbles. So ideally, the model parameters should only update after the actual/big batch_size. In pytorch, we did `.backward()` after each mini_batch to accumulate gradients, and `.step()` after each ppo_epoch.

Reply:
Thanks Ellen! I referred to the TRL impl, which does gradient update for each minibatch: https://github.com/huggingface/trl/blob/v0.7.4/trl/trainer/ppo_trainer.py#L761-L800

My understanding was that batch_size is set to be larger than mini_batch_size to parallelize and speed up rollouts. It's fine to me to do either way (apply gradient update per batch or per mini_batch).

In addition, in my experiments I always set batch_size == minibatch_size :)